File: brainfuck.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const VERSION = "v0.5"

const RESET = "\033[0m"
const RED = "\033[31m"
const GREEN = "\033[32m"
const GREY = "\033[90m"

func contains(array []rune, character rune) bool {
	for _, element := range array {
		if element == character {
			return true
		}
	}
	return false
}

func parse_input(input string) string {
	output := ""
	allowedCharacter := []rune{'>', '<', '+', '-', '.', ',', '[', ']'}

	for _, line := range strings.Split(input, "\n") {
		if line == "" {
			continue
		} else if line[0] != '#' {
			for _, character := range line {
				if character == '#' {
					break
				} else if contains(allowedCharacter, character) {
					output += string(character)
				}
			}
		}
	}

	return output
}

func interpret(input string) int {
	tape := [30000]byte{}
	tapePtr := 0
	inputPtr := 0

	for inputPtr < len(input) {
		switch input[inputPtr] {
		case '>':
			if tapePtr >= len(tape)-1 {
				tapePtr = 0
			} else {
				tapePtr++
			}
		case '<':
			if tapePtr <= 0 {
				tapePtr = len(tape) - 1
			} else {
				tapePtr--
			}
		case '+':
			if tape[tapePtr] >= 255 {
				tape[tapePtr] = 0
			} else {
				tape[tapePtr]++
			}
		case '-':
			if tape[tapePtr] <= 0 {
				tape[tapePtr] = 255
			} else {
				tape[tapePtr]--
			}
		case '.':
			fmt.Print(string(tape[tapePtr]))
		case ',':
			fmt.Print("Type a character: ")
			reader := bufio.NewReader(os.Stdin)
			characater, error := reader.ReadByte()

			if error != nil {
				fmt.Print(RED + "ERROR" + RESET + ": Cannot read the key\n")
				return 1
			}

			tape[tapePtr] = characater
		case '[':
			if tape[tapePtr] == 0 {
				inputPtrTemp := inputPtr
				count := 1

				for count != 0 {
					if inputPtrTemp >= len(input)-1 {
						fmt.Print(RED + "ERROR" + RESET + ": Missing closing bracket\n")
						return 1
					} else {
						inputPtrTemp++

						if input[inputPtrTemp] == '[' {
							count++
						} else if input[inputPtrTemp] == ']' {
							count--
						}
					}
				}

				inputPtr = inputPtrTemp
			}
		case ']':
			if tape[tapePtr] != 0 {
				inputPtrTemp := inputPtr
				count := 1

				for count != 0 {
					if inputPtrTemp <= 0 {
						fmt.Print(RED + "ERROR" + RESET + ": Missing opening bracket\n")
						return 1
					} else {
						inputPtrTemp--

						if input[inputPtrTemp] == '[' {
							count--
						} else if input[inputPtrTemp] == ']' {
							count++
						}
					}
				}

				inputPtr = inputPtrTemp
			}
		}

		inputPtr++
	}

	return 0
}

func main() {
	args := os.Args[1:]
	argc := len(args)

	if argc <= 0 {
		fmt.Print(
			GREEN+"Brainfuck interpreter\n"+RESET,
			"\n",
			GREY+"USAGE"+RESET+": brainfuck your_file.b\n",
			"\n",
			"--interpret -i     Input directly a string to interpret\n",
			"--version   -v     Print the interpreter version\n",
			"--github    -g     Give the GitHub link\n",
		)
	} else {
		switch args[0] {
		case "--version", "-v":
			fmt.Println(VERSION)
		case "--github", "-g":
			fmt.Print("https://github.com/Gael-Lopes-Da-Silva/Brainfuck\n")
		case "--interpret", "-i":
			os.Exit(interpret(parse_input(args[1])))
		default:
			fileContent, error := os.ReadFile(args[0])
			if error == nil {
				os.Exit(interpret(parse_input(string(fileContent))))
			} else {
				fmt.Print(RED + "ERROR" + RESET + ": File not found\n")
			}
		}
	}
}

File: Source/brainfuck.cs
/* 
@author: Gael Lopes Da Silva
@project: Brainfuck Interpreter
@github: https://github.com/Gael-Lopes-Da-Silva/Brainfuck
@gitlab: https://gitlab.com/Gael-Lopes-Da-Silva/Brainfuck
*/

public class Brainfuck
{
    private static void WriteColoredLine(string type, ConsoleColor color, string message)
    {
        Console.ForegroundColor = color;
        Console.Write(type.ToUpper());
        Console.ResetColor();
        Console.WriteLine(message);
    }

    private static string ParseInput(string[] input)
    {
        string parsedInput = "";
        char[] allowedCharacter = { '>', '<', '+', '-', '.', ',', '[', ']' };

        foreach (string line in input)
        {
            if (line == "")
            {
                continue;
            }
            else if (line[0] != '#')
            {
                foreach (char character in line)
                {
                    if (character == '#')
                    {
                        break;
                    }
                    else if (allowedCharacter.Contains(character))
                    {
                        parsedInput += character;
                    }
                }
            }
        }

        return parsedInput;
    }

    private static void Interpret(string input)
    {
        byte[] tape = new byte[30000];

        int tapePtr = 0;
        int inputPtr = 0;

        while (inputPtr < input.Length)
        {
            switch (input[inputPtr])
            {
                case '>':
                    if (tapePtr >= tape.Length - 1)
                    {
                        tapePtr = 0;
                    }
                    else
                    {
                        ++tapePtr;
                    }
                    break;

                case '<':
                    if (tapePtr <= 0)
                    {
                        tapePtr = tape.Length - 1;
                    }
                    else
                    {
                        --tapePtr;
                    }
                    break;

                case '+':
                    if (tape[tapePtr] >= 255)
                    {
                        tape[tapePtr] = 0;
                    }
                    else
                    {
                        tape[tapePtr]++;
                    }
                    break;

                case '-':
                    if (tape[tapePtr] <= 0)
                    {
                        tape[tapePtr] = 255;
                    }
                    else
                    {
                        tape[tapePtr]--;
                    }
                    break;

                case '.':
                    Console.Write((char)tape[tapePtr]);
                    break;

                case ',':
                    Console.Write("\nEnter a character: ");
                    tape[tapePtr] = (byte)Console.ReadKey().KeyChar;
                    Console.Write("\n");
                    break;

                case '[':
                    if (tape[tapePtr] == 0)
                    {
                        int inputPtrTemp = inputPtr;
                        int count = 1;

                        while (count != 0)
                        {
                            if (inputPtrTemp >= input.Length - 1)
                            {
                                WriteColoredLine("ERROR", ConsoleColor.Red, ": Missing closing bracket");
                                return;
                            }
                            else
                            {
                                inputPtrTemp++;

                                if (input[inputPtrTemp] == '[')
                                {
                                    count++;
                                }
                                else if (input[inputPtrTemp] == ']')
                                {
                                    count--;
                                }
                            }
                        }

                        inputPtr = inputPtrTemp;
                    }
                    break;

                case ']':
                    if (tape[tapePtr] != 0)
                    {
                        int inputPtrTemp = inputPtr;
                        int count = 1;

                        while (count != 0)
                        {
                            if (inputPtrTemp <= 0)
                            {
                                WriteColoredLine("ERROR", ConsoleColor.Red, ": Missing opening bracket");
                                return;
                            }
                            else
                            {
                                inputPtrTemp--;

                                if (input[inputPtrTemp] == '[')
                                {
                                    count--;
                                }
                                else if (input[inputPtrTemp] == ']')
                                {
                                    count++;
                                }
                            }
                        }

                        inputPtr = inputPtrTemp;
                    }
                    break;
            }

            inputPtr++;
        }
    }

    public static void Main(string[] args)
    {
        if (args.Length <= 0)
        {
            Console.WriteLine("Brainfuck Interpreter\n");
            WriteColoredLine("USAGE", ConsoleColor.DarkGray, ": brainfuck.exe [<your file>]\n");

            WriteColoredLine("--version  -v", ConsoleColor.Green, ": Display version");
            WriteColoredLine("--github   -g", ConsoleColor.Green, ": Give the github link");
        }
        else if (args.Length >= 1)
        {
            switch (args[0])
            {
                case "--version":
                case "-v":
                    Console.WriteLine("v0.4");
                    break;

                case "--github":
                case "-g":
                    Console.WriteLine("https://github.com/Gael-Lopes-Da-Silva/Brainfuck");
                    break;

                default:
                    try
                    {
                        string[] input = File.ReadAllLines(args[0]);
                        Interpret(ParseInput(input));
                    }
                    catch (FileNotFoundException)
                    {
                        WriteColoredLine("ERROR", ConsoleColor.Red, $": Cannot open '{args[0]}'");
                    }
                    break;
            }
        }
    }
}
